### PR TITLE
First pass at LinearRegression

### PR DESCRIPTION
**Pull request by dcrescim**
Here's a first pass at LinearRegression in scikit.js.

First some notes on how I'm approaching this.

### 1. Obviously we are keeping the signature as close to scikit-learn as possible.

I've done all of the methods except for "score". I've also added a function called "importModel" 
which does the "importing from python" that we talked about in the discussion here.

The LinearRegression in Scikit-Learn doesn't use gradient descent (like this one). It uses some 
production level linalg solvers. Eventually in the long run, we should do the same thing. This is a holdover
until we get some better / faster code in here. 

### 2. I think it's probable wise to use get/set methods to match the Scikit-Learn Api

For LinearRegression in sklearn, you can just go **lr.coef_** and **lr.intercept_** to get those values.
We store the internal model as tf.sequential, so it makes sense to basically use getters/setters to not
break the contract.

### 3. I'm not sure what our plan is with prettier.js? 
 
Currently I'm operating off of the .prettier file in the repo, but some of the code in the repo is ignoring that. Moreover, my personal preference is indentation is 2, and no semi-colons. Though, I don't really care that much. I more just wanna make sure we are all on the same page, and that it runs semi-automatically in the repo.

### 4. I've never used JSDoc before but it seems great. 

I'm gonna add it in my next pass. Just wanted to get this PR in before I go to bed. I also plan on adding a lot more tests.

### 5. I drunkenly asked my buddy to make a logo for scikit.js. 
I've attached it. If ya like it, we can start there. Otherwise, we can make something else.

Screen Shot 2021-09-29 at 10 25 51 PM

I've attached a png screenshot only because Github complains when I try to put in an svg :)

### CODE REVIEW

**Comment by risenW** [comment]
@dcrescim This

> Here's a first pass at LinearRegression in scikit.js.
> 
> First some notes on how I'm approaching this.
> 
> ### 1. Obviously we are keeping the signature as close to scikit-learn as possible.
> I've done all of the methods except for "score". I've also added a function called "importModel" which does the "importing from python" that we talked about in the discussion [here.](https://github.com/opensource9ja/scikit.js/discussions/4#discussioncomment-1400404)
> 
> The LinearRegression in Scikit-Learn doesn't use gradient descent (like this one). It uses some production level linalg solvers. Eventually in the long run, we should do the same thing. This is a holdover until we get some better / faster code in here.
> 
> ### 2. I think it's probably wise to use get/set methods to match the Scikit-Learn Api
> For LinearRegression in sklearn, you can just go **lr.coef_** and **lr.intercept_** to get those values. We store the internal model as tf.sequential, so it makes sense to basically use getters/setters to not break the contract.
> 
> ### 3. I'm not sure what our plan is with prettier.js?
> Currently I'm operating off of the .prettier file in the repo, but some of the code in the repo is ignoring that. Moreover, my personal preference is indentation is 2, and no semi-colons. Though, I don't really care that much. I more just wanna make sure we are all on the same page, and that it runs semi-automatically in the repo.
> 
> ### 4. I've never used JSDoc before but it seems great.
> I'm gonna add it in my next pass. Just wanted to get this PR in before I go to bed. I also plan on adding a lot more tests.
> 
> ### 5. I drunkenly asked my buddy to make a logo for scikit.js.
> I've attached it. If ya like it, we can start there. Otherwise, we can make something else.
> 
> Screen Shot 2021-09-29 at 10 25 51 PM
> 
> I've attached a png screenshot only because Github complains when I try to put in an svg :)

This is a great PR! I'll answer.comment on your approaches in same order

1. Yes, we want to model the Scikit-learn API as closely as possible.
2. Yes, let's use the getter/setters as well.
3. My bad, I probably didn't configure my VS code prettier properly. But yes, we should definitely all use the prettier in the repo
4. Awesome!
5. Looks great, and yes, we can definitely iterate on this going forward. 

Overall, great PR, once it's all ready, do tag me so I can do some review. Looking forward to your tests as well.

**Comment by risenW** [comment]
@dcrescim Just to flag that you should create an issue for this PR. So we can track what you're working on, and not duplicate work.

**Comment by dcrescim** [comment]
> @dcrescim Just to flag that you should create an issue for this PR. So we can track what you're working on, and not duplicate work.

Makes total sense. Here's the issue I [created](https://github.com/opensource9ja/scikit.js/issues/7). Feel free to write down your thoughts on possible solutions as well.

**Comment by risenW** [comment]
> I updated this PR to include basic implementations of Lasso, Ridge, and ElasicNet regressions. They use the same base class (SGD) to do all the heavy lifting. A couple thoughts on future directions / ideas.
> 
> 1. I think we should use the analytic solutions for LinearRegression, and RidgeRegression, and only use SGD for the other linear models. That's because those 2 algorithms have exact solutions. And if we use them, the fit procedure will be wayyyy faster (as well as not requiring an async function as the "fit" method.
> 2. If there aren't exact solutions then we should use SGD but with "smart" parameter defaults, ie... For logistic regression set default coefficients to 0. For Lasso / ElasicNet, set the intercept equal to the average of the y response variable.
> 3. I'm honestly not sure if we should normalize the inputs before we send them into the model. Certainly it would help convergence but I'm wondering if it breaks the "mental model" of the user. Usually people do normalization outside of their model, and their is no reason to do it twice. Maybe we should pass **normalize** as an argument but I'm torn on the best way forward here.

Hey @dcrescim Here are my thoughts 

Regarding questions 1 and 2, I think it would be better if we support both approaches, so users can specify which approach they want to use. We don't have to implement both from atm, but it's something you should think about in the API design.

For question 3, as you said, we should not normalize inputs, users should do that, so we are consistent with Scikit API.
